### PR TITLE
RATIS-1154. Add sync flag when write stream data

### PR DESCRIPTION
**Pull request by runzhiwang**
## What changes were proposed in this pull request?

Add sync flag when write stream data

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/RATIS-1154

## How was this patch tested?

new assert

### CODE REVIEW

File: ratis-test/src/test/java/org/apache/ratis/datastream/DataStreamBaseTest.java [hunk 1]
```diff
@@ -105,8 +106,14 @@ SingleDataStream getSingleDataStream(long callId) {
   static class SingleDataStream implements DataStream {
     private int byteWritten = 0;
     private final RaftClientRequest writeRequest;
+    private boolean isForced = false;
```
Comment (szetszwo) [inline ratis-test/src/test/java/org/apache/ratis/datastream/DataStreamBaseTest.java:109]: Let's recorded the forced position.
```
int forcedPosition = 0;
```

File: ratis-client/src/main/java/org/apache/ratis/client/api/DataStreamOutput.java [hunk 2]
```diff
@@ -26,5 +26,5 @@
 /** An asynchronous output stream supporting zero buffer copying. */
 public interface DataStreamOutput extends CloseAsync<DataStreamReply> {
   /** Send out the data in the buffer asynchronously */
-  CompletableFuture<DataStreamReply> writeAsync(ByteBuffer buf);
+  CompletableFuture<DataStreamReply> writeAsync(ByteBuffer buf, boolean sync);
```
Comment (szetszwo) [inline ratis-client/src/main/java/org/apache/ratis/client/api/DataStreamOutput.java:29]: We should add one more method
```suggestion
  default CompletableFuture<DataStreamReply> writeAsync(ByteBuffer buffer) {
    return writeAsync(buffer, false);
  }

  CompletableFuture<DataStreamReply> writeAsync(ByteBuffer buffer, boolean sync);
```

**Review by szetszwo — COMMENTED** [review]
The change looks good.  Just some minor comments.

**Review by szetszwo — APPROVED** [review]
+1 the change looks good.

**Comment by runzhiwang** [comment]
@szetszwo Thanks for review. Because CI passed, so I merge it.

**Pull request merged**